### ISSUE TITLE
Bad performance using local attention

### ISSUE THREAD

**Issue by MarcusLoppe** [lucidrains/meshgpt-pytorch#21]
I noticed that the local attention layers isn't that great. 
Below is a testing using 6 unique 3d mesh chairs that is duplicated 200 times = 1200 examples per epoch.
I've seen this pattern before but I wanted to test it on some more complex 3D mesh before posting a issue.

![bild](https://github.com/lucidrains/meshgpt-pytorch/assets/65302107/a0e66dfe-94d3-43b7-8f03-5151d2711f9c)
 

Another topic since the discord link isn't working: 
I did some testing with text using 10 mesh chairs, I apply augmentation so each chair got 3 variations.
Then i duplicated each variation 500 times so the total dataset size is 3000.
I encoded the meshes using text as well but just using the same word 'chair', but this proves that the text generation works.

After 22 minutes training the encoder (0.24 loss) and then 2.5 hrs training the transformer (0.0048) I got the result below. 
I trained the transformer on different learning rates but in total there was 30 epochs e.g 30x 3000= 90 000 steps.

Training using text is about x2 times slower so might be good to see if any improvements on that front can be made.
  
I provided the text "chair" and looped the generation to use different temperature values from 0 to 1.0 with 0.1 as stepping value.

![bild](https://github.com/lucidrains/meshgpt-pytorch/assets/65302107/92e2ddff-4a93-4f1b-97d1-1d315936e7e5)
![bild](https://github.com/lucidrains/meshgpt-pytorch/assets/65302107/10eba874-b651-41dc-80fd-c984a09dd8af)

**Comment by lucidrains** [comment]
@MarcusLoppe looking more and more promising! thank you Marcus!

also thanks for sharing your local attention result; as a last ditch attempt, could you retry on smaller window size, say 16 or 32? if that still doesn't yield good results, i'll concede and default the layers to 0

**Comment by lucidrains** [comment]
@MarcusLoppe for efficient text training, my plan is to introduce some wrapper class that accepts [this function](https://github.com/lucidrains/meshgpt-pytorch/blob/main/meshgpt_pytorch/meshgpt_pytorch.py#L1030) and does some caching of the strings to a memmapped file during the first epoch

**Comment by lucidrains** [comment]
@MarcusLoppe will probably get the text stuff done mid next week

**Comment by MarcusLoppe** [comment]
> @MarcusLoppe for efficient text training, my plan is to introduce some wrapper class that accepts [this function](https://github.com/lucidrains/meshgpt-pytorch/blob/main/meshgpt_pytorch/meshgpt_pytorch.py#L1030) and does some caching of the strings to a memmapped file during the first epoch. the trainer class can then prioritize preference of `text_embeds` over `texts`

Okay, well I tried to preprocess the text using the embed_texts function and create the the text_embeds dict key, but when I was training it was giving me some issue about backward gradient(?) was already used(?) when I applied the loss.

**Comment by lucidrains** [comment]
@MarcusLoppe oh, want to try 0.3.17? honestly didn't expect anyone to start doing text conditioned training this soon (this was not in the paper)

**Comment by MarcusLoppe** [comment]
> @MarcusLoppe looking more and more promising! thank you Marcus!
> 
> also thanks for sharing your local attention result; as a last ditch attempt, could you retry on smaller window size, say 16 or 32? if that still doesn't yield good results, i'll concede and default the layers to 0

Don't give up! :) 

I tried with local_attn_window_size set to 16 & 32, it wasn't better.

Have you given it any idea of implementing it in the SageConv instead? It's quite bare and "normal", I'm not sure if anyone tried GCN + Attention. 

![bild](https://github.com/lucidrains/meshgpt-pytorch/assets/65302107/48a49eba-dd7c-4121-bd04-93a031ce0768)

**Comment by lucidrains** [comment]
@MarcusLoppe ok, final attempt, maybe only for decoder? next week we can try [an attention variant for GNN](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.GAT.html)

as you can tell, i have a lot of faith..

**Comment by MarcusLoppe** [comment]
> @MarcusLoppe ok, final attempt, maybe only for decoder? next week we can try [an attention variant for GNN](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.GAT.html) in the encoder
> 
> as you can tell, i have a lot of faith..
> 
> signing off, have a good sunday!

Sad :/  I tried with a big dataset of 116 models with 5 permutations that totals 16900 items per epoch.
Little worse performance, It might work on the really big dataset but this is my results for now.

![bild](https://github.com/lucidrains/meshgpt-pytorch/assets/65302107/568f9741-1d87-4d30-a47a-85de4be7ef6c)

**Comment by lucidrains** [comment]
> > @MarcusLoppe ok, final attempt, maybe only for decoder? next week we can try [an attention variant for GNN](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.GAT.html) in the encoder
> > as you can tell, i have a lot of faith..
> > signing off, have a good sunday!
> 
> Sad :/ I tried with a big dataset of 116 models with 5 permutations that totals 16900 items per epoch. Little worse performance, It might work on the really big dataset but this is my results for now.
> 
> Edit: I did some quick GAT testing as well, I think SageConv might be better.
> 
> ![bild](https://private-user-images.githubusercontent.com/65302107/291125561-d7017d54-9fff-45b7-acde-c6400c9f816b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MTcwMTgsIm5iZiI6MTcwMjkxNjcxOCwicGF0aCI6Ii82NTMwMjEwNy8yOTExMjU1NjEtZDcwMTdkNTQtOWZmZi00NWI3LWFjZGUtYzY0MDBjOWY4MTZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE2MjUxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU4YmI3Yjc1YjllOTVlN2FhYjBlNWUzMDYyZDY0ZGU4ZjY5NDJlNTExNmRiNWVjZWNjZDg4ZTBlODhlNzY3NjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QBqw5Ed6msHtl5ia73Ng_2GdraBGMXPQuN9CzprKt4c)

ahh thanks for the data point

i'll default the local attn layers to 0 for now until i give it a try

**Comment by MarcusLoppe** [comment]
> Hi, could you provide a sample code for the training with the text? Thanks in advance !

 I just create the dataset using a dict, then the trainer will pass it appropriately.

```
obj_data = {"vertices": torch.tensor(item["vertices"], dtype=torch.float), "faces":  torch.tensor(item["faces"], dtype=torch.long),"texts": item["texts"] } 
obj_datas.append(obj_data)


class MeshDataset(Dataset): 
    
    def __init__(self, data): 
        self.data = data
        print(f"Got {len(data)} data")

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx): 
        return self.data[idx]
```

**Comment by MarcusLoppe** [comment]
> @MarcusLoppe So, I was testing your notebook, I am using Colab instead of kaggle, how do I work with this: tables = load_json("/kaggle/input/shapenet/data.json",2), I have tried to search this json file online, but couldn't find it.


Oh I used shapenet for the 3D mesh, I can't give the data since ToS but take your own data and load using e.g load_models and then dump the json. 

The load_json will just load the data and convert the vertices/faces to tensors.

```
tables = load_models(r" filtered",5,5)  
with open("data.json", "w") as json_file:
    json.dump(tables, json_file)
```

**Comment by lucidrains** [comment]
local attention layers are defaulted to 0
